File: Dynamic_Programming/1696.Jump_Game_VI/Python/1696.Jump_Game_VI.py
class Solution(object):
    def maxResult(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        #============================================#
        # Dynamic-programming based traversal method #
        #============================================#

        ############
        #Initialize
        ##### Length of nums array #####
        len_nums = len(nums)

        ##### Record dynamic-programming array #####
        record_dp = [0 for _ in range(len_nums)]

        record_dp[0] = nums[0]

        ##### Record priority-queue #####
        record_pq = [[(record_dp[0] * (-1)), 0]]

        heapify(record_pq)


        ##########################################
        #Dynamic-programming based loop traversal
        for nums_idx in range(1, len_nums):
            while (record_pq and ((nums_idx - (record_pq[0])[1]) > k)):
                heappop(record_pq) #Keep updating/popped
            
            record_dp[nums_idx] += (((record_pq[0])[0] * (-1)) + nums[nums_idx]) #Keep updating/overwriting

            heappush(record_pq, [(record_dp[nums_idx] * (-1)), nums_idx]) #Keep updating/recording

        return record_dp[(len_nums - 1)]

File: Dynamic_Programming/1696.Jump_Game_VI/C++/1696.Jump_Game_VI.cpp
class Solution {
public:
    int maxResult(vector<int> & nums, int k) 
    {
        //====================================//
        // Input type:                        //
        //   - vector<int> ref. (nums vector) //
        //   - int (k-th steps)               //
        // Return type:                       //
        //   - int (result maximun score)     //
        //====================================//

        /*Initialize*/
        ///// Length of nums vector /////
        int len_nums = nums.size();

        ///// Record dynamic-programming (i.e. DP) vector /////
        vector<int> record_dp(len_nums, 0);

        record_dp[0] = nums[0];

        ///// Record priority-queue /////
        priority_queue<pair<int, int>> record_pq;

        record_pq.push(make_pair(record_dp[0], 0));


        /*Dynamic-programming based loop traversal*/
        for (int nums_idx = 1; (nums_idx < len_nums); (nums_idx++)) //Whole
        {
            while ((! record_pq.empty()) && ((nums_idx - (record_pq.top()).second) > k)) //Part
            {
                record_pq.pop(); //Keep updating, popped

            } //Part

            record_dp[nums_idx] += (((record_pq.top()).first) + nums[nums_idx]); //Keep updating, overwriting

            record_pq.push(make_pair(record_dp[nums_idx], nums_idx)); //Keep updating, recording

        } //Whole

        return record_dp[(len_nums - 1)];

    }
};
